subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchedHierarchies.java

```diff
@@ -16,99 +16,58 @@
-     */
-    @VisibleForTesting
-    static Set<Path> resolveHierarchiesToWatch(Stream<Path> directories) {
-        Set<Path> hierarchies = new HashSet<>();
-        directories
-            .sorted(Comparator.comparingInt(Path::getNameCount))
-            .filter(path -> {
-                Path parent = path;
-                while (true) {
-                    parent = parent.getParent();
-                    if (parent == null) {
-                        break;
-                    }
-                    if (hierarchies.contains(parent)) {
-                        return false;
-                    }
-                }
-                return true;
-            })
-            .forEach(hierarchies::add);
-        return hierarchies;
+    public Set<File> getWatchedRoots() {
+        return watchedRoots;
     }
 
-    private static class FilterAlreadyCoveredSnapshotsVisitor implements SnapshotHierarchy.SnapshotVisitor {
-        private final SnapshotHierarchy.SnapshotVisitor delegate;
-        private final FileHierarchySet alreadyCoveredSnapshots;
-
-        public FilterAlreadyCoveredSnapshotsVisitor(SnapshotHierarchy.SnapshotVisitor delegate, FileHierarchySet alreadyCoveredSnapshots) {
-            this.delegate = delegate;
-            this.alreadyCoveredSnapshots = alreadyCoveredSnapshots;
-        }
+    public static WatchedHierarchies resolveWatchedHierarchies(WatchableHierarchies watchableHierarchies, SnapshotHierarchy vfsRoot) {
+        FileHierarchySet watchedHierarchies = DefaultFileHierarchySet.of();
+        for (File watchableHierarchy : watchableHierarchies.getRecentlyUsedHierarchies()) {
+            String watchableHierarchyPath = watchableHierarchy.getAbsolutePath();
+            if (watchedHierarchies.contains(watchableHierarchyPath)) {
+                continue;
+            }
 
-        @Override
-        public void visitSnapshotRoot(FileSystemLocationSnapshot snapshot) {
-            if (!alreadyCoveredSnapshots.contains(snapshot.getAbsolutePath())) {
-                delegate.visitSnapshotRoot(snapshot);
+            if (hasNoContent(vfsRoot.rootSnapshotsUnder(watchableHierarchyPath), watchableHierarchies, watchedHierarchies)) {
+                continue;
             }
+            watchedHierarchies = watchedHierarchies.plus(watchableHierarchy);
         }
+        ImmutableSet.Builder<File> roots = ImmutableSet.builder();
+        watchedHierarchies.visitRoots(root -> roots.add(new File(root)));
+        return new WatchedHierarchies(watchedHierarchies, roots.build());
+    }
+
+    private static boolean hasNoContent(Stream<FileSystemLocationSnapshot> snapshots, WatchableHierarchies watchableHierarchies, FileHierarchySet watchedHierarchies) {
+        return snapshots
+            .filter(snapshot -> !watchedHierarchies.contains(snapshot.getAbsolutePath()))
+            .noneMatch(snapshot -> snapshot.getType() != FileType.Missing
```

Comment:
Do you mean we should _only_ check for `DIRECT` access?

Reply:
No. We should check that we only have missing snapshots with access type `DIRECT`.  I realize that my suggestion up there is wrong, since we are searching for an existing snapshot. So it should be
```
(snapshot.getType() != FileType.Missing  || snapshot.getAccessType() != FileMetadata.AccessType.DIRECT)
```

Reply:
I find it hard to understand the `noneMatch` condition with all the negations. How about using
```java
allMatch(snapshot -> (snapshot.getType() == FileType.Missing && snapshot.getAccessType() == DIRECT)
  || watchableHierarchies.ignoredForWatching(snapshot)) 
```

Reply:
I've addressed this now.